client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java

```diff
@@ -242,36 +248,37 @@ private boolean skipLocation(int startMapIndex, int endMapIndex, PartitionLocati
       return true;
     }
 
-    private PartitionLocation nextReadableLocation() {
-      int locationCount = locations.length;
+    private Tuple2<PartitionLocation, PbStreamHandler> nextReadableLocation() {
+      int locationCount = locations.size();
       if (fileIndex >= locationCount) {
         return null;
       }
-      PartitionLocation currentLocation = locations[fileIndex];
+      PartitionLocation currentLocation = locations.get(fileIndex);
       while (skipLocation(startMapIndex, endMapIndex, currentLocation)) {
         skipCount.increment();
         fileIndex++;
         if (fileIndex == locationCount) {
           return null;
         }
-        currentLocation = locations[fileIndex];
+        currentLocation = locations.get(fileIndex);
       }
 
       fetchChunkRetryCnt = 0;
 
-      return currentLocation;
+      return new Tuple2(
+          currentLocation, streamHandlers == null ? null : streamHandlers.get(fileIndex));
```

Comment:
Actually we can't remove because `streamHandlers` is ArrayList and the elements in each position matches the element in `locations`, and we use `fileIndex` to index into the array.